src/lib/peers/PeerManager.ts

```diff
@@ -236,80 +201,92 @@ class PeerManager {
-      try {
-        tasks.push(socialDiscovery.findUser(handle, service));
-      } catch (error) {
-        console.log(`Could not find user on this discovery service: ${socialDiscovery.name}`);
-      }
-    }
+  async findSocialUser(handle: string, service: string, timeout?: number): Promise<boolean> {
+    const tasks = this.socialDiscoveryServices.map((s) => s.findUser(handle, service));
 
-    const pubKeyOrFail = await firstPromiseFulfilled(tasks);
-    if (pubKeyOrFail.length > 1) {
+    const pubKeyOrFail = await Promise.race(tasks);
+
+    if (!pubKeyOrFail) {
       throw Error('Could not find public key from services');
     }
 
-    const pubKeyFound = pubKeyOrFail[0];
-    const peerInfo = await this.findPubKey(pubKeyFound);
-    return peerInfo;
+    return await this.findPublicKey(pubKeyOrFail, timeout);
   }
 
   ///////////////////////
   // Peers Connections //
   ///////////////////////
   /**
    * Get a secure connection to the peer
-   * @param peer Public key of an existing peer or address of new peer
+   * @param publicKey Public key of an existing peer or address of new peer
    */
-  connectToPeer(peer: string | Address): GitFrontend {
+  connectToPeer(publicKey: string): PeerConnection {
     // Throw error if trying to connect to self
-    if (peer == this.localPeerInfo.connectedAddr || peer == this.localPeerInfo.publicKey) {
+    if (publicKey == this.peerInfo.publicKey) {
       throw Error('Cannot connect to self');
     }
-    let address: Address;
-    if (typeof peer == 'string') {
-      const existingConnection = this.peerConnections.get(peer);
-      if (existingConnection) {
-        return existingConnection;
-      }
-      const peerAddress = this.getPeer(peer)?.connectedAddr;
 
-      if (peerAddress) {
-        address = peerAddress;
-      } else {
-        throw Error('Peer does not exist in peer store');
-      }
-    } else {
-      address = peer;
+    const existingSocket = this.peerConnections.get(publicKey);
+    if (existingSocket) {
+      return existingSocket;
     }
 
-    const conn = new GitFrontend(address, this.keyManager);
+    // try to create a connection to the address
+    const peerConnection = new PeerConnection(publicKey, this.keyManager, this);
 
-    if (typeof peer == 'string') {
-      this.peerConnections.set(peer, conn);
-    }
+    this.peerConnections.set(publicKey, peerConnection);
+
+    return peerConnection;
+  }
 
-    return conn;
+  async pingPeer(publicKey: string, timeout?: number): Promise<boolean> {
+    const peerConnection = this.connectToPeer(publicKey);
+    return await peerConnection.pingPeer(timeout);
   }
 
   /* ============ HELPERS =============== */
   private writeMetadata(): void {
-    const peerInfo = this.localPeerInfo;
-    const metadata = peer.PeerInfoMessage.encode({
-      addresses: peerInfo.AdressStringList,
-      connectedAddr: peerInfo.connectedAddr?.toString(),
-      pubKey: peerInfo.publicKey,
+    // write peer info
```

Comment:
Any plans to use DHT?

Reply:
yeah definitely, I think that will be an enhancement for peer discovery across multiple hops for nodes that don't use any kind of social discovery service (or perhaps for discovery of machine nodes?)

Reply:
for now though, I think social discovery and nat traversal is enough purely because we most likely won't have large, federated hierarchies for a while. I will add an issue and label it enhancement :+1:

Reply:
https://github.com/MatrixAI/js-polykey/issues/97